src/run_program.rs

```diff
@@ -326,18 +326,89 @@ pub fn run_program<T: Allocator + 'static>(
 
 #[test]
 fn test_msb_mask() {
-    assert!(msb_mask(0x0) == 0x0);
-    assert!(msb_mask(0x01) == 0x01);
-    assert!(msb_mask(0x02) == 0x02);
-    assert!(msb_mask(0x04) == 0x04);
-    assert!(msb_mask(0x08) == 0x08);
-    assert!(msb_mask(0x10) == 0x10);
-    assert!(msb_mask(0x20) == 0x20);
-    assert!(msb_mask(0x40) == 0x40);
-    assert!(msb_mask(0x80) == 0x80);
-
-    assert!(msb_mask(0x44) == 0x40);
-    assert!(msb_mask(0x2a) == 0x20);
-    assert!(msb_mask(0xff) == 0x80);
-    assert!(msb_mask(0x0f) == 0x08);
+    assert_eq!(msb_mask(0x0), 0x0);
+    assert_eq!(msb_mask(0x01), 0x01);
+    assert_eq!(msb_mask(0x02), 0x02);
+    assert_eq!(msb_mask(0x04), 0x04);
+    assert_eq!(msb_mask(0x08), 0x08);
+    assert_eq!(msb_mask(0x10), 0x10);
+    assert_eq!(msb_mask(0x20), 0x20);
+    assert_eq!(msb_mask(0x40), 0x40);
+    assert_eq!(msb_mask(0x80), 0x80);
+
+    assert_eq!(msb_mask(0x44), 0x40);
+    assert_eq!(msb_mask(0x2a), 0x20);
+    assert_eq!(msb_mask(0xff), 0x80);
+    assert_eq!(msb_mask(0x0f), 0x08);
+}
+
+#[test]
+fn test_first_non_zero() {
+    assert_eq!(first_non_zero(&[]), 0);
+    assert_eq!(first_non_zero(&[1]), 0);
+    assert_eq!(first_non_zero(&[0]), 1);
+    assert_eq!(first_non_zero(&[0, 0, 0, 1, 1, 1]), 3);
+    assert_eq!(first_non_zero(&[0, 0, 0, 0, 0, 0]), 6);
+    assert_eq!(first_non_zero(&[1, 0, 0, 0, 0, 0]), 0);
+}
+
+#[test]
+fn test_traverse_path() {
+    use crate::int_allocator::IntAllocator;
+
+    let a = IntAllocator::new();
+    let nul = a.null();
+    let n1 = a.new_atom(&[0, 1, 2]);
+    let n2 = a.new_atom(&[4, 5, 6]);
+
+    assert_eq!(traverse_path(&a, &[0], &n1).unwrap(), Reduction(2, nul));
```

Comment:
it's a little bit unintuitive that this should cost 2, but I'm pretty sure this is what the python code does too, because this counts both as one leading zero byte as well as the base cost of 1